### PR TITLE
CB-18823 - add s3guard validation

### PR DESCRIPTION
**Pull request by vivprabu**
See detailed description in the commit message.

### CODE REVIEW

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/validation/handler/ClusterUpgradeS3guardValidationHandler.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+    @Override
+    public String selector() {
+        return VALIDATE_S3GUARD_DISABLED_EVENT.selector();
+    }
+
+    @Override
+    protected Selectable defaultFailureEvent(Long resourceId, Exception e, Event<ClusterUpgradeS3guardValidationEvent> event) {
+        LOGGER.error("Cluster upgrade S3guard validation was unsuccessful due to an unexpected error", e);
+        return new ClusterUpgradeS3guardValidationFinishedEvent(START_CLUSTER_UPGRADE_IMAGE_VALIDATION_EVENT.name(),
```
Comment (lturcsanyi) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/validation/handler/ClusterUpgradeS3guardValidationHandler.java:44]: Please move the selector into the constructuor.

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/validation/handler/ClusterUpgradeS3guardValidationHandler.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+                throw new UpgradeValidationFailedException("S3Guard is enabled. " +
+                    "Please disable it by following: https://docs.cloudera.com/cdp-public-cloud-preview-features/cloud/disable-s3-guard/disable-s3-guard.pdf");
+            }
+            return new ClusterUpgradeS3guardValidationEvent(getNextSelector(), request.getResourceId(), request.getImageId());
+        } catch (UpgradeValidationFailedException ex) {
+            LOGGER.warn("Cluster upgrade validation failed", ex);
+            return new ClusterUpgradeValidationFailureEvent(stackId, ex);
+        } catch (Exception ex) {
+            LOGGER.error("Cluster upgrade validation was unsuccessful due to an internal error", ex);
+            return new ClusterUpgradeS3guardValidationEvent(getNextSelector(), request.getResourceId(), request.getImageId());
```
Comment (lturcsanyi) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/validation/handler/ClusterUpgradeS3guardValidationHandler.java:69]: It should be `new ClusterUpgradeS3guardValidationFinishedEvent(...)`

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/validation/handler/ClusterUpgradeS3guardValidationHandler.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            String environmentCrn = stackService.findEnvironmentCrnByStackId(stackId);
+            DetailedEnvironmentResponse environmentResponse = environmentService.getByCrn(environmentCrn);
+            boolean s3guardEnabled = environmentResponse.getAws() != null && environmentResponse.getAws().getS3guard() != null
+                    && environmentResponse.getAws().getS3guard().getDynamoDbTableName() != null;
+            if (s3guardEnabled) {
+                LOGGER.error("Upgrade validation failed for resource: {} as S3Guard is enabled", request.getResourceId());
+                throw new UpgradeValidationFailedException("S3Guard is enabled. " +
+                    "Please disable it by following: https://docs.cloudera.com/cdp-public-cloud-preview-features/cloud/disable-s3-guard/disable-s3-guard.pdf");
+            }
+            return new ClusterUpgradeS3guardValidationEvent(getNextSelector(), request.getResourceId(), request.getImageId());
```
Comment (lturcsanyi) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/validation/handler/ClusterUpgradeS3guardValidationHandler.java:63]: It should be `new ClusterUpgradeS3guardValidationFinishedEvent(...)`

File: core/src/test/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/validation/handler/ClusterUpgradeS3guardValidationHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    private StackService stackService;
+
+    @Mock
+    private EnvironmentClientService environmentService;
+
+    private DetailedEnvironmentResponse detailedEnvironmentResponse;
+
+    @Test
+    public void testHandlerToCheckNonS3guardEnvironment() {
+        when(stackService.findEnvironmentCrnByStackId(STACK_ID)).thenReturn("1");
```
Comment (gergopapi2) [inline core/src/test/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/validation/handler/ClusterUpgradeS3guardValidationHandlerTest.java:45]: It makes sense to give name to constants.

Comment (gergopapi2) [reply core/src/test/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/validation/handler/ClusterUpgradeS3guardValidationHandlerTest.java:45]: Also, please add tests for the selector and defaultFailureEvent.

**Review by akanto — CHANGES_REQUESTED** [review]
As it was explained in the jira ticket, this should not be in CB but in CM, therefore this pull request can't be approved or merged.

**Comment by keyki** [comment]
As discussed, the PR will be merged once the commands are addressed.

**Comment by akanto** [comment]
> As it was explained in the jira ticket, this should not be in CB but in CM, therefore this pull request can't be approved or merged.

We have clarified with Shwetha, the validation part can be merged, however we need to make sure that the code does not have any CM-related interaction feature (e.g CM API call) or anything that should be done in CM upgrade handler.

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/validation/handler/ClusterUpgradeS3guardValidationHandler.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        Long stackId = request.getResourceId();
+        try {
+            String environmentCrn = stackService.findEnvironmentCrnByStackId(stackId);
+            DetailedEnvironmentResponse environmentResponse = environmentService.getByCrn(environmentCrn);
+            boolean s3guardEnabled = environmentResponse.getAws() != null && environmentResponse.getAws().getS3guard() != null
+                    && environmentResponse.getAws().getS3guard().getDynamoDbTableName() != null;
+            if (s3guardEnabled) {
+                LOGGER.error("Upgrade validation failed for resource: {} as S3Guard is enabled", request.getResourceId());
+                return new ClusterUpgradeValidationFailureEvent(stackId, new UpgradeValidationFailedException("S3Guard is enabled. " +
+                "Please disable it by following: https://docs.cloudera.com/cdp-public-cloud-preview-features/cloud/disable-s3-guard/disable-s3-guard.pdf"));
```
[the remark this answers is not in the document]
Comment (vivprabu) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/datalake/upgrade/validation/handler/ClusterUpgradeS3guardValidationHandler.java]: done

**Review by gergopapi2 — APPROVED** [review]

**Review by lturcsanyi — APPROVED** [review]

**Review by gergopapi2 — APPROVED** [review]

**Review by akanto — APPROVED** [review]

**Review by doktoric — APPROVED** [review]

**Pull request merged**